### ISSUE TITLE
partx cannot remove /dev/sdb3 on the origin disk (usb) during migration

### ISSUE THREAD

**Issue by dmuhamedagic** [drakkar-lig/debootstick#22]
The migration went good, but then there was this message:
```
partx: /dev/sdb: error deleting partition 3
```
BTW, why is it that you want to remove the installation from the USB disk?

**Comment by eduble** [comment]
> BTW, why is it that you want to remove the installation from the USB disk?

That is what debootstick does, when using migration feature: it _moves_ the OS, it _does not copy_ it.

This LVM-based technique has some advantages:
- It moves the OS while it's running. You do not even need a reboot after that.
- You can redo it later. Imagine that later, in a few years, you buy a faster disk with higher capacity: you can move the OS again. With this technique the OS can be seen as a "nomadic thing" actually.

But, since the OS is moved, it is no longer present on your initial disk or USB device. (the code that failed asks the kernel to unregister partitions, because you probably want to unplug the USB device at this time.)
Thus this technique is limited to single-machine installation.

At this time, debootstick is not suited to mass-installation, if ever this is what you expected.

There might be other options in the future, but for now I do not have a clear idea of how to achieve this.

The main issue is that new LVM versions needs the OS to be booted to work properly. And once the OS is booted, it is probably a wrong idea to just _copy_ it, because some running processes may have created or modified files. Because of that, booting the OS copied this way on the new device may fail with unexpected issues.

Using an LVM snapshot to record the state of the OS before it is started could be an option. But this would force debootstick to reserve room for this LVM snapshot. How much? That is the question. debootstick also has to reserve room at the filesystem level, so the trade off is hard to find.

**Comment by dmuhamedagic** [comment]
I didn't actually know that pvmove(8) is used. Yes, maybe I'll take a look at the LVM snapshot, but that's not that important now.

**Comment by dmuhamedagic** [comment]
Trying again partx -d /dev/sdb succeeds. What do you make of that? Some timing issue with dm or LVM?

**Comment by eduble** [comment]
Probably. The other issue you mentioned, about the target disk, can effectively occur since I forgot the target disk could be previously configured with LVM.
But, for this original disk, we know how it is structured and I believe the script does what it should do to release it.

This kind of code could probably do the trick:

```
while true
do
   sleep 0.1
   partx -d ${ORIGIN} || continue
   break
done
```

We have to be careful: using `partx -d ${ORIGIN} && break` would actually cause the script to exit right away if `partx` fails,  because of the `set -e` flag at the top of the script.

**Comment by dmuhamedagic** [comment]
I'm a bit under pressure right now, so I just commented out that part. Anyway, do you think it's necessary to remove the partitions too? Because, anyway, the vg and the lv were removed, who cares if the partition table remains.

**Comment by eduble** [comment]
> I'm a bit under pressure right now, so I just commented out that part.

I see, same here...

> Anyway, do you think it's necessary to remove the partitions too? Because, anyway, the vg and the lv were removed, who cares if the partition table remains.

Sure, it is not that important. It's just to be clean and avoid any kernel warning if the user unplugs the USB device after migration.

**Comment by dmuhamedagic** [comment]
I implemented a fix according to your comment (with retries), just need to clean it up a bit.

**Comment by eduble** [comment]
This is also fixed by commit f22b47d5c07e989d1dae019606943e82420a5240, adapted from your code.
